### ISSUE TITLE
[Hierarchy] Ability to set several Objects as a Unified Render Group

### ISSUE THREAD

**Issue by OverboyDev** [Scirra/Construct-feature-requests#19]
### Reviewed guidelines

- [X] I have read and understand the suggestion guidelines

### Checked for duplicate suggestions

- [X] I checked for existing similar suggestions

### Summary

![RenderGroup (1)](https://github.com/Scirra/Construct-feature-requests/assets/99751718/34e53233-98f4-4ff5-b33b-769aa39b23c1)

![294931243-008b0f1a-f571-4b00-9d02-76250db5a42f](https://github.com/Scirra/Construct-feature-requests/assets/99751718/eb8176fd-5146-4c67-a556-691e60edea1a)

### Possible workarounds or alternatives

There is no way to have several Objects acting like a single "Merged Object Render" in Construct.
Right now, we can set-up a Hierarchy made of multiple objects but each of those objects is rendered seperately. Which means for example, even if the children have the Opacity boolean set to true, it acts as several semi-transparent objects on top of each other (see Mockup1).
It's also a problem for effects. Even if each object of the Hierarchy has a outline effect for example, the outline effect would be applied individually on each Object.

Another issue is that while it's great that a Hierarchy can be made of Objects with totally different ZOrder and even different layers, it sometimes painful to make sure the whole Hierarchy or just a part of it is always rendered at the same ZOrder.

For example, let's say I have a topdown 2D game where characters are Skeletal 2D hierarchies. It's a nightmare to handle the ZOrder of those Hierarchies.

Same for many more things such as blending Mode.

**Which is why a "Unifed Render Group" Hierarchy setting.**

### Proposed solution

When adding an Object as Child to a Parent Object, **add a "Render Group" boolean option in Hierarchy Children Settings.**

(Let's call the Object at the top of a Render Group a "Render Leader" or a "Leader" for the explanation.)

If Render group is set to true, then :

- the Child Layer is locked to **the Leader Layer**
**- The ZOrder of this object is now obligatory adjacent to the other Objects that are part of this Render Group** (this involves several things at editor and runtime).
     - Basically only the Leader Object at the top of the "Render Group" can change its actual Zorder at edittime and at runtime, and all the children objects part of this "render group" are automatically moved at the right ZOrder.
     - Children part of a "Render Group" can only be moved at the top/at the bottom of their Render Group, or moved under/above an Object that's also part of this Render Group
- The Object at the top of the Render Group applies its Layers, Visibility, Opacity, ZOrder, Blending Mode and effects to the whole Render Group uniformly. It almost acts like a "mini-layer". 

**NOTES :**

- A render Group can be just a part of a Hierarchy , the "Render Group Leader" (parent) can be an object that isn't the top parent of the whole Hierarchy.
- The smallest render groups you can imagine is an Object (Parent) with a Hiearchy Child with "Render Group" boolean set to true.
     - If the Child has children (grandchildren of Parent), only the grandchildren with "Render Group" boolean set to true are part of the Render Group, the other just acts as regular Hierarchy children of those objects.
     - If a great children has its "Render Group" boolean set to false but has a children (so great grandchildren of the original Parent) with a "Render Group" boolean set to true. Then the grandchildren is the "Leader" of a new "Render Group" with that great grandchildren part of it. 
- (there is 2 Render Groups inside the same hierarchy)
- If someday a Hierarchy View is added, it could display the Render Group in an intuitive way **(see mockup 2)**
- The render group text in parenthesis are just explanations, they wouldn't be actually displayed,
- The Render Group line could be displayed only when an optional "Display All render Groups" is true or only if an object part of this Render Group is selected

### Why is this idea important?

Would allow to set the opacity or ZOrder or apply effects on a group of objects as if they were only 1 object
(useful for composite objects)

**Comment by WilsonPercival** [comment]
Workaround: 
[issue19_workaround.zip](https://github.com/Scirra/Construct-feature-requests/files/13897611/issue19_workaround.zip)

**Comment by OverboyDev** [comment]
Yeah this is interesting, I think Scirra could use a similar trick under the hood to implement such a feature now dynamic layers are a thing.

We would still need an actual editor feature for it so it's useable in production and it works well with everything else in the engine.

**Comment by WilsonPercival** [comment]
> Yeah this is interesting, I think Scirra could use a similar trick under the hood to implement such a feature now dynamic layers are a thing.
> 
> We would still need an actual editor feature for it so it's useable in production and it works well with everything else in the engine.

@OverboyDev In that case, if I were you, I would write what you mean, because in the current situation it looks like your problem has been completely resolved.

**Comment by OverboyDev** [comment]
No it isn't. I would be able to create dynamic layers for my "render groups", but let's say all or even just a few of the entities of my topdown game are composite objects, then how would I handle the Zordering of my entities if half of them are just regular objects and some other are dynamic layers ?

It would be messy because some of my "instances" would be actual object instance while other would be in fact layers. 
So i wouldn't be able to share common logic between them, like a Family would allow to do, same for effects. 

One benefit of this feature request would be that the workflow would be exactly the same as regular objects for composite objects. You would just set the Zorder / Opacity / Effects on the "RenderGroup Leader" and Construct would handle it pleasantly for us.

Also it would be just a checkbox to set in the Hierarchy settings while the editor setup would be awful if we tried to implement this ourself.
This is just a few reasons why this would be a nightmare, there are many more.

Also if Hierarchy View is added someday, it could display render group pleasantly as in my mockup.

![294931243-008b0f1a-f571-4b00-9d02-76250db5a42f](https://github.com/Scirra/Construct-feature-requests/assets/99751718/5d63c47d-dd3c-4156-a55a-24abf67f6876)

**Comment by WilsonPercival** [comment]
@OverboyDev In this case, you could create each instance on its own dynamic layer. You could then move layers around, sorting them by the Y coordinate of the instance on that layer.

**Comment by XHXIAIEIN** [comment]
Yes, I have the same problem.
![357833209-272b4537-8987-42d2-9448-c756fec9f064](https://github.com/user-attachments/assets/6e190ac8-0957-429a-9519-73f655a634a5)

Thanks to @WilsonPercival for the Workaround, I got the inspiration and we can indeed do this with layers.

[UsingLayerGroupMultiChildObjectOpacity.c3p.zip](https://github.com/user-attachments/files/16615195/UsingLayerGroupMultiChildObjectOpacity.c3p.zip)

It would be nice if the hierarchy supported an option to help us do this more easily.
